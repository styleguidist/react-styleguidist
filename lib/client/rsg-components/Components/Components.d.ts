import React from 'react';
interface ComponentsProps {
    components: any[];
    depth: number;
    exampleMode?: string;
    usageMode?: string;
}
declare const Components: React.FunctionComponent<ComponentsProps>;
export default Components;
