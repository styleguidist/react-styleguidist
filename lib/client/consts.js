import "core-js/modules/es.object.freeze";
export var DisplayModes = Object.freeze({
  // Show all sections and components (default)
  all: 'all',
  // Show one section
  section: 'section',
  // Show one component
  component: 'component',
  // Show one example inside component or section
  example: 'example',
  // Show error 404
  notFound: 'notFound'
});
export var ExampleModes = Object.freeze({
  hide: 'hide',
  collapse: 'collapse',
  expand: 'expand'
});
export var UsageModes = Object.freeze({
  hide: 'hide',
  collapse: 'collapse',
  expand: 'expand'
});