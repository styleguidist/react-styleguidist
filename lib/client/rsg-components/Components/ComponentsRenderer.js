import React from 'react';
import PropTypes from 'prop-types';
export default function ComponentsRenderer(_ref) {
  var children = _ref.children;
  return /*#__PURE__*/React.createElement("div", null, children);
}
ComponentsRenderer.propTypes = {
  children: PropTypes.node.isRequired
};