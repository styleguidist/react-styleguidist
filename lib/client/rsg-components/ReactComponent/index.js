export { default } from 'rsg-components/ReactComponent/ReactComponent';
export * from 'rsg-components/ReactComponent/ReactComponent';