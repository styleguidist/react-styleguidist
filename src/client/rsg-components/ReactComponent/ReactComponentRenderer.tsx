import React from 'react';
import PropTypes from 'prop-types';
import Pathline from 'rsg-components/Pathline';
import Styled, { JssInjectedProps } from 'rsg-components/Styled';
import * as Rsg from '../../../typings';

const styles = ({ color, fontSize, space }: Rsg.Theme) => ({
	root: {
		marginBottom: space[6],
	},
	header: {
		marginBottom: space[3],
	},
	tabs: {
		marginBottom: space[3],
	},
	tabButtons: {
		marginBottom: space[1],
	},
	tabBody: {
		overflowX: 'auto',
		maxWidth: '100%',
		WebkitOverflowScrolling: 'touch',
	},
	docs: {
		color: color.base,
		fontSize: fontSize.text,
	},
});

interface ReactComponentRendererProps extends JssInjectedProps {
	name: string;
	heading: React.ReactNode;
	filepath?: string;
	slug?: string;
	pathLine?: string;
	tabButtons?: React.ReactNode;
	tabBody?: React.ReactNode;
	description?: React.ReactNode;
	docs?: React.ReactNode;
	examples?: React.ReactNode;
	isolated?: boolean;
}

export const ReactComponentRenderer: React.FunctionComponent<ReactComponentRendererProps> = ({
	classes,
	name,
	heading,
	pathLine,
	description,
	docs,
	examples,
	tabButtons,
	tabBody,
}) => {
	return (
		<div className={classes.root} data-testid={`${name}-container`}>
			<header className={classes.header}>
				{heading}
				{pathLine && <Pathline>{pathLine}</Pathline>}
			</header>
			{(description || docs) && (
				<div className={classes.docs}>
					{description}
					{docs}
				</div>
			)}
			{tabButtons && (
				<div className={classes.tabs}>
					<div className={classes.tabButtons}>{tabButtons}</div>
					<div className={classes.tabBody}>{tabBody}</div>
				</div>
			)}
			{examples}
		</div>
	);
};

ReactComponentRenderer.propTypes = {
	classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	name: PropTypes.string.isRequired,
	heading: PropTypes.any.isRequired,
	filepath: PropTypes.string,
	pathLine: PropTypes.string,
	tabButtons: PropTypes.any,
	tabBody: PropTypes.any,
	description: PropTypes.any,
	docs: PropTypes.any,
	examples: PropTypes.any,
	isolated: PropTypes.bool,
};

export default Styled<ReactComponentRendererProps>(styles)(ReactComponentRenderer);
