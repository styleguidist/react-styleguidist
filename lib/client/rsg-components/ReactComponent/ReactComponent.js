import "core-js/modules/es.symbol";
import "core-js/modules/es.symbol.description";
import "core-js/modules/es.array.concat";
import "core-js/modules/es.function.name";
import "core-js/modules/es.object.assign";
import "core-js/modules/es.object.get-prototype-of";
import "core-js/modules/es.object.to-string";
import "core-js/modules/es.reflect.construct";
import "core-js/modules/es.regexp.to-string";

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Examples from 'rsg-components/Examples';
import SectionHeading from 'rsg-components/SectionHeading';
import JsDoc from 'rsg-components/JsDoc';
import Markdown from 'rsg-components/Markdown';
import Slot from 'rsg-components/Slot';
import ReactComponentRenderer from 'rsg-components/ReactComponent/ReactComponentRenderer';
import Context from 'rsg-components/Context';
import ExamplePlaceholderDefault from 'rsg-components/ExamplePlaceholder';
import { DOCS_TAB_USAGE } from '../slots';
import { DisplayModes, UsageModes } from '../../consts';
var ExamplePlaceholder = process.env.STYLEGUIDIST_ENV !== 'production' ? ExamplePlaceholderDefault : function () {
  return /*#__PURE__*/React.createElement("div", null);
};

var ReactComponent = /*#__PURE__*/function (_Component) {
  _inheritsLoose(ReactComponent, _Component);

  var _super = _createSuper(ReactComponent);

  function ReactComponent() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_this), "state", {
      activeTab: _this.props.usageMode === UsageModes.expand ? DOCS_TAB_USAGE : undefined
    });

    _defineProperty(_assertThisInitialized(_this), "handleTabChange", function (name) {
      _this.setState(function (state) {
        return {
          activeTab: state.activeTab !== name ? name : undefined
        };
      });
    });

    return _this;
  }

  var _proto = ReactComponent.prototype;

  _proto.render = function render() {
    var activeTab = this.state.activeTab;
    var _this$context = this.context,
        displayMode = _this$context.displayMode,
        pagePerSection = _this$context.config.pagePerSection;
    var _this$props = this.props,
        component = _this$props.component,
        depth = _this$props.depth,
        usageMode = _this$props.usageMode,
        exampleMode = _this$props.exampleMode;
    var name = component.name,
        visibleName = component.visibleName,
        _component$slug = component.slug,
        slug = _component$slug === void 0 ? '-' : _component$slug,
        filepath = component.filepath,
        pathLine = component.pathLine,
        href = component.href;

    var _ref = component.props || {},
        _ref$description = _ref.description,
        description = _ref$description === void 0 ? '' : _ref$description,
        _ref$examples = _ref.examples,
        examples = _ref$examples === void 0 ? [] : _ref$examples,
        _ref$tags = _ref.tags,
        tags = _ref$tags === void 0 ? {} : _ref$tags;

    if (!name) {
      return null;
    }

    var showUsage = usageMode !== UsageModes.hide;
    return /*#__PURE__*/React.createElement(ReactComponentRenderer, {
      name: name,
      slug: slug,
      filepath: filepath,
      pathLine: pathLine,
      docs: /*#__PURE__*/React.createElement(JsDoc, tags),
      description: description && /*#__PURE__*/React.createElement(Markdown, {
        text: description
      }),
      heading: /*#__PURE__*/React.createElement(SectionHeading, {
        id: slug,
        pagePerSection: pagePerSection,
        deprecated: !!tags.deprecated,
        slotName: "componentToolbar",
        slotProps: Object.assign({}, component, {
          isolated: displayMode !== DisplayModes.all
        }),
        href: href,
        depth: depth
      }, visibleName),
      examples: examples.length > 0 ? /*#__PURE__*/React.createElement(Examples, {
        examples: examples,
        name: name,
        exampleMode: exampleMode
      }) : /*#__PURE__*/React.createElement(ExamplePlaceholder, {
        name: name
      }),
      tabButtons: showUsage && /*#__PURE__*/React.createElement(Slot, {
        name: "docsTabButtons",
        active: activeTab,
        props: Object.assign({}, component, {
          onClick: this.handleTabChange
        })
      }),
      tabBody: /*#__PURE__*/React.createElement(Slot, {
        name: "docsTabs",
        active: activeTab,
        onlyActive: true,
        props: component
      })
    });
  };

  return ReactComponent;
}(Component);

_defineProperty(ReactComponent, "propTypes", {
  component: PropTypes.object.isRequired,
  depth: PropTypes.number.isRequired,
  exampleMode: PropTypes.string.isRequired,
  usageMode: PropTypes.string.isRequired
});

_defineProperty(ReactComponent, "contextType", Context);

export { ReactComponent as default };