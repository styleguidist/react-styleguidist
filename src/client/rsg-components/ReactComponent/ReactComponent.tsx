import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Examples from 'rsg-components/Examples';
import SectionHeading from 'rsg-components/SectionHeading';
import JsDoc from 'rsg-components/JsDoc';
import Markdown from 'rsg-components/Markdown';
import Slot from 'rsg-components/Slot';
import ReactComponentRenderer from 'rsg-components/ReactComponent/ReactComponentRenderer';
import Context, { StyleGuideContextContents } from 'rsg-components/Context';
import ExamplePlaceholderDefault from 'rsg-components/ExamplePlaceholder';
import { DOCS_TAB_USAGE } from '../slots';
import { DisplayModes, UsageModes } from '../../consts';
import * as Rsg from '../../../typings';

const ExamplePlaceholder =
	process.env.STYLEGUIDIST_ENV !== 'production' ? ExamplePlaceholderDefault : () => <div />;

interface ReactComponentProps {
	component: Rsg.Component;
	depth: number;
	exampleMode?: string;
	usageMode?: string;
}

interface ReactComponentState {
	activeTab?: string;
}

export default class ReactComponent extends Component<ReactComponentProps, ReactComponentState> {
	public static propTypes = {
		component: PropTypes.object.isRequired,
		depth: PropTypes.number.isRequired,
		exampleMode: PropTypes.string.isRequired,
		usageMode: PropTypes.string.isRequired,
	};

	public static contextType = Context;

	public state = {
		activeTab: this.props.usageMode === UsageModes.expand ? DOCS_TAB_USAGE : undefined,
	};

	private handleTabChange = (name: string) => {
		this.setState((state) => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	};

	public render() {
		const { activeTab } = this.state;
		const {
			displayMode,
			config: { pagePerSection },
		} = this.context as StyleGuideContextContents;
		const { component, depth, usageMode, exampleMode } = this.props;
		const { name, visibleName, slug = '-', filepath, pathLine, href } = component;
		const { description = '', examples = [], tags = {} } = component.props || {};
		if (!name) {
			return null;
		}
		const showUsage = usageMode !== UsageModes.hide;

		return (
			<ReactComponentRenderer
				name={name}
				slug={slug}
				filepath={filepath}
				pathLine={pathLine}
				docs={<JsDoc {...tags} />}
				description={description && <Markdown text={description} />}
				heading={
					<SectionHeading
						id={slug}
						pagePerSection={pagePerSection}
						deprecated={!!tags.deprecated}
						slotName="componentToolbar"
						slotProps={{
							...component,
							isolated: displayMode !== DisplayModes.all,
						}}
						href={href}
						depth={depth}
					>
						{visibleName}
					</SectionHeading>
				}
				examples={
					examples.length > 0 ? (
						<Examples examples={examples} name={name} exampleMode={exampleMode} />
					) : (
						<ExamplePlaceholder name={name} />
					)
				}
				tabButtons={
					showUsage && (
						<Slot
							name="docsTabButtons"
							active={activeTab}
							props={{ ...component, onClick: this.handleTabChange }}
						/>
					)
				}
				tabBody={<Slot name="docsTabs" active={activeTab} onlyActive props={component} />}
			/>
		);
	}
}
