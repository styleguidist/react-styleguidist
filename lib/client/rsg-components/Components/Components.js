import "core-js/modules/es.array.map";
import React from 'react';
import PropTypes from 'prop-types';
import ReactComponent from 'rsg-components/ReactComponent';
import ComponentsRenderer from 'rsg-components/Components/ComponentsRenderer';

var Components = function Components(_ref) {
  var components = _ref.components,
      depth = _ref.depth,
      exampleMode = _ref.exampleMode,
      usageMode = _ref.usageMode;
  return /*#__PURE__*/React.createElement(ComponentsRenderer, null, components.map(function (component) {
    return /*#__PURE__*/React.createElement(ReactComponent, {
      key: component.filepath,
      component: component,
      exampleMode: exampleMode,
      usageMode: usageMode,
      depth: depth
    });
  }));
};

Components.propTypes = {
  components: PropTypes.array.isRequired,
  depth: PropTypes.number.isRequired,
  exampleMode: PropTypes.string.isRequired,
  usageMode: PropTypes.string.isRequired
};
export default Components;