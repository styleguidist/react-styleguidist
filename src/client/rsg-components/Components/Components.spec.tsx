import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import ReactComponent from '../ReactComponent';
import Components from './Components';
import ComponentsRenderer from './ComponentsRenderer';
import { ExampleModes, UsageModes } from '../../consts';

const exampleMode = ExampleModes.collapse;
const usageMode = UsageModes.collapse;
const components = [
	{
		name: 'Foo',
		pathLine: 'components/foo.js',
		filepath: 'components/foo.js',
		props: {
			description: 'Foo foo',
		},
	},
	{
		name: 'Bar',
		pathLine: 'components/bar.js',
		filepath: 'components/bar.js',
		props: {
			description: 'Bar bar',
		},
	},
];

it('should render components list', () => {
	const renderer = createRenderer();
	renderer.render(
		<Components components={components} exampleMode={exampleMode} usageMode={usageMode} depth={3} />
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('renderer should render components list layout', () => {
	const renderer = createRenderer();
	renderer.render(
		<ComponentsRenderer>
			<ReactComponent
				key={0}
				component={components[0]}
				exampleMode={exampleMode}
				usageMode={usageMode}
				depth={3}
			/>
			<ReactComponent
				key={1}
				component={components[1]}
				exampleMode={exampleMode}
				usageMode={usageMode}
				depth={3}
			/>
		</ComponentsRenderer>
	);

	expect(renderer.getRenderOutput()).toMatchSnapshot();
});
