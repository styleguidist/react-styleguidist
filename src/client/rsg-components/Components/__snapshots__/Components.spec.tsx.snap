// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render components list layout 1`] = `
<div>
  <ReactComponent
    component={
      Object {
        "filepath": "components/foo.js",
        "name": "Foo",
        "pathLine": "components/foo.js",
        "props": Object {
          "description": "Foo foo",
        },
      }
    }
    depth={3}
    exampleMode="collapse"
    usageMode="collapse"
  />
  <ReactComponent
    component={
      Object {
        "filepath": "components/bar.js",
        "name": "Bar",
        "pathLine": "components/bar.js",
        "props": Object {
          "description": "Bar bar",
        },
      }
    }
    depth={3}
    exampleMode="collapse"
    usageMode="collapse"
  />
</div>
`;

exports[`should render components list 1`] = `
<ComponentsRenderer>
  <ReactComponent
    component={
      Object {
        "filepath": "components/foo.js",
        "name": "Foo",
        "pathLine": "components/foo.js",
        "props": Object {
          "description": "Foo foo",
        },
      }
    }
    depth={3}
    exampleMode="collapse"
    usageMode="collapse"
  />
  <ReactComponent
    component={
      Object {
        "filepath": "components/bar.js",
        "name": "Bar",
        "pathLine": "components/bar.js",
        "props": Object {
          "description": "Bar bar",
        },
      }
    }
    depth={3}
    exampleMode="collapse"
    usageMode="collapse"
  />
</ComponentsRenderer>
`;
